### PR TITLE
Feature/time triggered batch

### PR DESCRIPTION
**Pull request by HubertBubert**

### CODE REVIEW

File: quedex_api/user_stream.py [hunk 1]
```diff
@@ -185,6 +261,22 @@ class UserStream(object):
   class - see their comments for the format of the data.
   """
 
+  class BatchMode(Enum):
+    STANDARD = 1
+    TIME_TRIGGERED_CREATE = 2
+    TIME_TRIGGERED_UPDATE = 3
+
+  type_listener  = {
```
Comment (wgromniak) [inline quedex_api/user_stream.py]: Variable name could be better. Not sure what this is.

File: quedex_api/user_stream.py [hunk 2]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
     """
     pass
 
+  def on_time_triggered_batch_added(self, time_triggered_batch_added):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_added(self, timer_added)

File: quedex_api/user_stream.py [hunk 3]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_added: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_rejected(self, time_triggered_batch_rejected):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_rejected(self, timer_rejected)

File: quedex_api/user_stream.py [hunk 4]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_rejected: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+        "cause": "too_many_active_timers"/"timer_already_expired"/"timer_already_exists",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_expired(self, time_triggered_batch_expired):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_expired(self, timer_expired)

File: quedex_api/user_stream.py [hunk 5]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_expired: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_triggered(self, time_triggered_batch_triggered):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_triggered(self, timer_triggerred)

File: quedex_api/user_stream.py [hunk 6]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_triggered: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_updated(self, time_triggered_batch_updated):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_updated(self, timer_updated)

File: quedex_api/user_stream.py [hunk 7]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_updated: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_update_failed(self, time_triggered_batch_update_failed):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_update_failed(self, timer_update_failed)

File: quedex_api/user_stream.py [hunk 8]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_update_failed: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+        "cause": "not_found"/"timer_execution_interval_broken",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_cancelled(self, time_triggered_batch_cancelled):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_update_cancelled(self, timer_update_cancelled)

File: quedex_api/user_stream.py [hunk 9]
```diff
@@ -150,6 +151,81 @@ def on_order_filled(self, order_filled):
...
+    """
+    :param time_triggered_batch_cancelled: a dict of the following format:
+      {
+        "timer_id": "<string id>",
+      }
+    """
+    pass
+
+  def on_time_triggered_batch_cancel_failed(self, time_triggered_batch_cancel_failed):
```
Comment (mchalapuk) [inline quedex_api/user_stream.py]: on_timer_update_cancel_failed(self, timer_update_cancel_failed)

File: quedex_api/user_stream.py (hunk 1 above)
Comment (mchalapuk) [reply quedex_api/user_stream.py]: I'd say this is unnecessary. Please remove this.

File: quedex_api/user_stream.py [hunk 10]
```diff
@@ -308,51 +401,271 @@ def batch(self, order_commands):
...
-    self._batching = True
+    self._batch_mode = self.BatchMode.STANDARD
 
   def send_batch(self):
     """
     Sends batch created from calling place_order, cancel_order, modify_order after calling
     start_batch.
     """
-    if not self._batching:
+    if not (self._batch_mode == self.BatchMode.STANDARD):
```
[the remark this answers is not in the document]
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Any suggestions?

File: quedex_api/user_stream.py [hunk 11]
```diff
@@ -308,51 +401,271 @@ def batch(self, order_commands):
...
     if len(self._batch) == 0:
       raise ValueError("Empty batch")
     self._send_batch_no_checks(self._batch)
     self._batch = None
-    self._batching = False
+    self._batch_mode = None
+
+  def time_triggered_batch(self, batch_id, execution_start_timestamp, execution_expiration_timestamp, order_commands):
+    """
+    Creates timer in exchange engine with a batch of order commands. These commands will be automatically
```
[the remark this answers is not in the document]
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Currently api allows to batch commands created by these methods only:
- place_order
- cancel_order
- cancel_all_orders
- modify_order
All of these are order commands.

File: quedex_api/user_stream.py [hunk 12]
```diff
@@ -308,51 +401,271 @@ def batch(self, order_commands):
...
+        "type": "modify_order",
+        // for the rest of the fields see modify_order method
+      },
+      {
+        "type": "cancel_all_orders",
+      },
+      ...
+     ]
+
+
```
[the remark this answers is not in the document]
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Removed.

File: tests/test_user_stream.py [hunk 13]
```diff
@@ -7,6 +7,8 @@
 
 
 class TestUserStream(TestCase):
+  maxDiff = None
```
[the remark this answers is not in the document]
Comment (HubertBubert) [reply tests/test_user_stream.py]: This allows printing a full diff on error.

File: quedex_api/user_stream.py (hunk 1 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Removed.

File: quedex_api/user_stream.py (hunk 2 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 3 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 4 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 6 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 7 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 8 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 9 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

File: quedex_api/user_stream.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (wgromniak) [reply quedex_api/user_stream.py]: `if self._batch_mode != self.BatchMode.STANDARD`?

File: quedex_api/user_stream.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wgromniak) [reply quedex_api/user_stream.py]: OK

File: tests/test_user_stream.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (wgromniak) [reply tests/test_user_stream.py]: OK

File: quedex_api/user_stream.py [hunk 14]
```diff
@@ -429,13 +418,21 @@ def send_batch(self):
...
+    Sends a time triggered batch with the given list of order commands to the exchange.
+    When a time triggered batch is received by the exchange engine, a new timer is registered.
+    Based on the timer configuration, at some point in the future (between executionStartTimestamp and executionExpirationTimestamp),
+    all the carried order commands are processed, one by one, in the creation order. If executionStartTimestamp is in the past
+    and executionExpirationTimestamp is in the future, the batch will be processed immediately.
+
+    Timers are ordered by executionStartTimestamp and order of arrival. Timers with earlier executionStartTimestamp are processed first.
+    When there are more timers with the same executionStartTimestamp they are triggered by the order of arrival.
+
+    Note: timers are evaluated each time a new command is processed by the exchange engine. If between executionStartTimestamp
```
Comment (wgromniak) [inline quedex_api/user_stream.py]: As in java - unnecessary implementation detail. And what's important from user perspective - min. granularity of execution and why expiration is relevant.

File: tests/test_user_stream.py [hunk 15]
```diff
@@ -7,7 +7,7 @@
 
 
 class TestUserStream(TestCase):
-  maxDiff = None
+  # maxDiff = None
```
Comment (wgromniak) [inline tests/test_user_stream.py]: Commented code.

File: quedex_api/user_stream.py (hunk 14 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Removed. Added link to native API documentation.

**Comment by HubertBubert** [comment]
Added links to the native API documentation.

File: quedex_api/user_stream.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (HubertBubert) [reply quedex_api/user_stream.py]: :) Modified.

File: quedex_api/user_stream.py (hunk 5 above)
Comment (HubertBubert) [reply quedex_api/user_stream.py]: Changed.

**Pull request merged**